src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11302,15 +11320,55 @@ private bool ScanCast(bool forPattern = false)
 
         private bool IsPossibleLambdaExpression(Precedence precedence)
         {
-            // Only call into this if after `static` or after a legal identifier.
+            // Only call into this if after `static`, '[',  or after a legal identifier.
             Debug.Assert(
-                this.CurrentToken.Kind == SyntaxKind.StaticKeyword ||
+                this.CurrentToken.Kind is SyntaxKind.StaticKeyword or SyntaxKind.OpenBracketToken ||
                 this.IsTrueIdentifier(this.CurrentToken));
             if (precedence > Precedence.Lambda)
             {
                 return false;
             }
 
+            if (this.CurrentToken.Kind == SyntaxKind.OpenBracketToken)
+            {
+                var resetPoint = this.GetResetPoint();
+
+                _ = ParseAttributeDeclarations();
+                bool result;
+
+                switch (CurrentToken.Kind)
+                {
+                    case SyntaxKind.StaticKeyword:
+                    case SyntaxKind.IdentifierToken:
+                        result = IsPossibleLambdaExpressionCore(precedence);
+                        break;
+                    case SyntaxKind.OpenParenToken:
+                        result = ScanParenthesizedImplicitlyTypedLambda(precedence) || ScanExplicitlyTypedLambda(precedence);
+                        break;
+                    default:
+                        result = false;
+                        break;
+                }
+
+                this.Reset(ref resetPoint);
+                this.Release(ref resetPoint);
```

Comment:
we generally put these in finallys.  is it ok that these aren't?

Reply:
also, if you do a try/finally, the body of this helper becomes much simpler,a nd less likely to have a problem later on (for example, if someone errantly returns).